### ISSUE TITLE
Dynamic `blockquote.twitter-tweet` embeds

### ISSUE THREAD

**Issue by CyberShadow** [zedeus/nitter#586]
It would be great to also support Twitter's dynamic embedding mechanism in conjunction with e.g. libredirect.

Two examples of pages that currently don't work:

- https://www.reddit.com/live/18hnzysb1elcs/
  - Has many iframes to redditmedia.com, for example: https://www.redditmedia.com/mediaembed/liveupdate/18hnzysb1elcs/LiveUpdate_112dbf2c-aad3-11ec-b1dc-9e39912881e9/0
- https://mashable.com/article/best-funniest-tweets-week-february-18

Both of these pages seem to use a similar mechanism:

- A `blockquote` with `class="twitter-tweet"` is present in the page HTML, containing some placeholder text and a link (`a` tag) to the tweet
- The page loads Twitter's `widgets.js` script: `<script async src="https://platform.twitter.com/widgets.js" charset="utf-8"></script>`
- This script searches for `blockquote.twitter-tweet` and turns it into Twitter embeds (iframes).

If Twitter's `widgets.js` is allowed to execute, it does create iframes which libredirect successfully redirects to Nitter embeds. However, these iframes remain hidden. I think `widgets.js` is waiting for the iframe to somehow signal that it finished loading, but that never happens in the Nitter implementation.

I can see one of two approaches to make this work:

- Make Nitter's iframes compatible with Twitter's `widgets.js` script
- Implement Nitter's own `widgets.js` script, which libredirect could load instead of the Twitter one.

Twitter's documentation for this mechanism:

- https://developer.twitter.com/en/docs/twitter-for-websites/javascript-api/guides/set-up-twitter-for-websites (how to add `widgets.js`)
- https://developer.twitter.com/en/docs/twitter-for-websites/embedded-tweets/overview (the "Render a Tweet with JavaScript" section - describes the `blockquote.twitter-tweet` convention)

Libredirect discussion: https://github.com/libredirect/libredirect/issues/20

**Comment by LainLayer** [comment]
Hi, this raises multiple issues.
Mainly, the idea of nitter is to be a more usable lighter and faster front end to twitter. If we do the same thing twitter does, then there is no point.
Second, we are trying to implement this using no JavaScript, to not slow down website that use nitter embeds, load faster, avoid content jumping and more.
This complicates things a bit, but we think its worth the effort, evidently since the mashable link you sent took me around a minute to load.

As to your problem, embeds are still a prototype, we are waiting on some fixes in other places to make them closer to actually usable for the purpose of replacing embeds on sites automatically. It will be a bit easier after that.

For now just replace the whole twitter blockquote with a nitter iframe, the height will be a bit weird but its something.

**Comment by catwithbanana** [comment]
I want to add my voice in hoping this issue is resolved. Nitter embeds + LibRedirect are the only way to view Twitter embeds while keeping Firefox's Strict Enhanced Tracking Protection turned on. Toggling it off for every website that relies on Twitter embeds means surfing a large portion of the internet without essential privacy features such as Total Cookie Protection.

Nitter embeds were working wonderfully until April 10, when they simply stopped working across all websites for all users simultaneously. [I filed an issue with LibRedirect](https://github.com/libredirect/libredirect/issues/183#issuecomment-1144111047) but after some back and forth was notified that any potential fix would have to happen to Nitter's codebase. As of now while using Firefox Enhanced Tracking Protection set to Strict one must click through every single Twitter embed individually to see what was said. It's not ideal.

I'm not a programmer or developer, so I can't deign to understand what exactly needs to be done, but I really hope this issue is resolved, as it is an absolutely killer feature. I'm even willing to contribute money towards a developer and/or bounty if that's what it takes to devise a durable solution for fixing Nitter embeds across most websites. Please let me know how I can help, and thanks for all of your hard work.

**Comment by Victor239** [comment]
@catwithbanana I think you need to file a separate issue. This one is referring to replacing embeds when the tweet has a non-standard height, as Nitter embeds don't dynamically adjust their height in relation to the tweet content. But now instead there's an issue where _every embed_ is no longer working.

**Comment by catwithbanana** [comment]
> @catwithbanana I think you need to file a separate issue. This one is referring to replacing embeds when the tweet has a non-standard height, as Nitter embeds don't dynamically adjust their height in relation to the tweet content. But now instead there's an issue where _every embed_ is no longer working.

Done -- thank you for the heads up.

**Comment by HookedBehemoth** [comment]
```js
let tweet_regex = /^https?:\/\/twitter\.com\/[^/]+\/status\/(\d+)\??/
let blockquotes = document.querySelectorAll("blockquote.twitter-tweet")
for (i = 0; i < blockquotes.length; ++i) {
    let blockquote = blockquotes.item(i)
    let link = blockquote.lastChild.href
    let id = tweet_regex.exec(link)[1]
    let embed = document.createElement("iframe")
    embed.src = `https://nitter.cunnycon.org/i/status/${id}/embed`
    embed.style = 'width:100%;height:400px'
    embed.loading = 'lazy'
    blockquote.parentNode.replaceChildren(embed)
}
```
I wrote this code a week back which works well enough as a userscript but isn't perfect.
No idea how you would determine the right height. You could also serve this as widgets.js.

Maybe this helps someone implement a better and more complete solution.

![image](https://user-images.githubusercontent.com/22580720/174103956-30a768ce-8a26-491b-b3f6-41140f58cc88.png)

**Comment by zedeus** [comment]
Modified the script slightly since it didn't work for me:
```js
let tweet_regex = /^https?:\/\/twitter\.com\/[^/]+\/status\/(\d+)\??/
let blockquotes = document.querySelectorAll("blockquote.twitter-tweet")
for (i = 0; i < blockquotes.length; ++i) {
    let blockquote = blockquotes[i]
    let link = blockquote.children[0].href
    let id = tweet_regex.exec(link)[1]
    let embed = document.createElement("iframe")
    embed.src = `https://nitter.cunnycon.org/i/status/${id}/embed`
    embed.style = 'width:100%;height:400px'
    embed.loading = 'lazy'
    blockquote.replaceChildren(embed)
}
```

**Comment by HookedBehemoth** [comment]
`blockquote.children[0].href` doesn't seem right.
The link is always the last one. finding an `a` element would probably better though.

**Comment by zedeus** [comment]
This works:

```js
let tweet_regex = /^https?:\/\/twitter\.com\/[^/]+\/status\/(\d+)\??/
let blockquotes = document.querySelectorAll("blockquote.twitter-tweet a")
for (i = 0; i < blockquotes.length; ++i) {
    let blockquote = blockquotes[i]
    let link = blockquote.href
    let id = tweet_regex.exec(link)[1]
    let embed = document.createElement("iframe")
    embed.src = `https://nitter.cunnycon.org/i/status/${id}/embed?theme=Twitter`
    embed.style = 'width:100%;height:400px'
    embed.loading = 'lazy'
    blockquote.replaceWith(embed)
}
```

![image](https://user-images.githubusercontent.com/9434921/174206739-16ff7d76-3411-456a-91d2-0cf117d9f2bc.png)

**Comment by HookedBehemoth** [comment]
That breaks if there are other links in the blockquote.

**Comment by zedeus** [comment]
`blockquote.twitter-tweet > a` should do it

**Comment by Victor239** [comment]
If you get a userscript working, may I suggest posting it on [OpenUserJS](https://openuserjs.org/) or [Greasy Fork](https://greasyfork.org) or a GitHub gist so that others can use it also.

**Comment by HookedBehemoth** [comment]
After a lot of back and forth I concluded that the only reliable way of doing this is:
- set iframe height to something large that would fit the entire tweet (optionally disable resizing)
- set 'load' listener inside the iframe
  - send message to parent with `window.parent.postMessage`
- listen in parent for this custom message

With 'load' awaiting full page load this works even without setting individual image sizes.

The big drawback is that this requires enabling javascript on the parent site.

https://github.com/zedeus/nitter/commit/06679c6f64427187a4dbaf48f6d24b35ea4bb13f
```js
let iframeMap = {}

const instance = "https://nitter.cunnycon.org"
const tweet_regex = /^https?:\/\/twitter\.com\/[^/]+\/status\/(\d+)\??/
const blockquotes = document.querySelectorAll("blockquote.twitter-tweet > a")
for (const blockquote of blockquotes) {
  const link = blockquote.href
  const id = tweet_regex.exec(link)[1]
  const embed = document.createElement("iframe")
  const url = `${instance}/i/status/${id}/embed`
  embed.src = url
  embed.style = 'width:100%;height:600px'
  embed.loading = 'lazy'
  embed.id = "target"
  blockquote.parentNode.replaceChildren(embed)
  if (iframeMap[url]) {
    iframeMap[url].push(embed)
  } else {
    iframeMap[url] = [embed]
  }
}

window.addEventListener('message', function(e) {
  if (e.origin != instance || e.data[0] != 'resizeIframe')
    return

  const data = e.data[1];
  for (const embed of iframeMap[data["url"]]) {
    embed.style.height = `${data["h"]}px`
  }
}, false);```

**Comment by HookedBehemoth** [comment]
Added the script to the PR, ready for review.
